### DISCUSSION TITLE
How to call a generic method?

### DISCUSSION THREAD

**Discussion by kajr0s** [vfsfitvnm/frida-il2cpp-bridge#466] [Q&A] [answered]
I want to call a method GameObject.GetComponent<T>, but

const image = obj.method("GetComponent").inflate(imageType).invoke();

returns 'number', not 'Il2Cpp.Object'.

namespace Verne
{
	// Token: 0x0200022E RID: 558
	[Token(Token = "0x2000148")]
	public class HelpController : MonoBehaviour
	{
		// Token: 0x06000D8E RID: 3470 RVA: 0x00002050 File Offset: 0x00000250
		[Token(Token = "0x60009B2")]
		[Address(RVA = "0xD9748C", Offset = "0xD9748C", VA = "0xD9748C")]
		private void OnHelpIconCreated(GameObject obj)
		{
		}

namespace UnityEngine
{
	// Token: 0x020000AE RID: 174
	[Token(Token = "0x20000AE")]
	[UsedByNativeCode]
	[Attribute(Name = "NativeHeaderAttribute", RVA = "0x8EDAF4", Offset = "0x8EDAF4")]
	[ExcludeFromPreset]
	public sealed class GameObject : Object
	{
		// Token: 0x0600044A RID: 1098 RVA: 0x00002052 File Offset: 0x00000252
		[Token(Token = "0x600044A")]
		public T GetComponent<T>()
		{
			return null;
		}

import "frida-il2cpp-bridge";

Il2Cpp.perform(() => {
    // Il2Cpp.dump();
    // console.log(Il2Cpp.unityVersion);

    const assemblyCSharp = Il2Cpp.domain.assembly("Assembly-CSharp").image;

    const helpController = assemblyCSharp.class("Verne.HelpController");
    const onHelpIconCreated = helpController.method("OnHelpIconCreated");
    const unityEngineUI = Il2Cpp.domain.assembly("UnityEngine.UI").image;
    const imageType = unityEngineUI.class("UnityEngine.UI.Image");

    // @ts-ignore
    onHelpIconCreated.implementation = function(this: Il2Cpp.Object, obj: Il2Cpp.Object) {
        // ??? obj.GetComponent<Image>() return 'number'
        const image = obj.method("GetComponent").inflate(imageType).invoke();
        if (image)
        {
            console.log("Component 'Image' found");
            const sprite = image.method("get_sprite").invoke();

C:/code/frida-agent-example/agent/index.ts (62,34): Property 'method' does not exist on type 'number | true | NativePointer | Int64 | UInt64 | Object | Pointer<Type> | ValueType | String | Array<Type
Property 'method' does not exist on type 'number'.
Error: compilation failed
    at Module.build (file:///C:/code/frida-agent-example/node_modules/frida-compile/dist/compiler.js:34:15)
    at main (file:///C:/code/frida-agent-example/node_modules/frida-compile/dist/cli.js:37:33)
    at file:///C:/code/frida-agent-example/node_modules/frida-compile/dist/cli.js:57:5
    at ModuleJob.run (node:internal/modules/esm/module_job:198:25)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:385:24)
    at async loadESM (node:internal/process/esm_loader:88:5)
    at async handleMainPromise (node:internal/modules/run_main:61:12)

**Comment by kajr0s** [comment]
Alternatively, call another overload method 'public extern Component GetComponent(Type type)'

	// Token: 0x020000AE RID: 174
	[Token(Token = "0x20000AE")]
	[UsedByNativeCode]
	[Attribute(Name = "NativeHeaderAttribute", RVA = "0x8EDAF4", Offset = "0x8EDAF4")]
	[ExcludeFromPreset]
	public sealed class GameObject : Object
	{
		// Token: 0x0600044B RID: 1099
		[Token(Token = "0x600044B")]
		[Address(RVA = "0x150BFA8", Offset = "0x150BFA8", VA = "0x150BFA8")]
		[Attribute(Name = "FreeFunctionAttribute", RVA = "0x8F4AC8", Offset = "0x8F4AC8")]
		[Attribute(Name = "TypeInferenceRuleAttribute", RVA = "0x8F4AC8", Offset = "0x8F4AC8")]
		[MethodImpl(4096)]
		public extern Component GetComponent(Type type);

but I don’t understand how to pass the type in the parameter.

**Comment by thinhbuzz** [comment, marked as answer]
Use imageType.type.object

**Reply by kajr0s** [reply]
Thanks, this really works!!!
